mesonbuild/rewriter.py

```diff
@@ -633,6 +635,166 @@ def find_assignment_node(self, node: BaseNode) -> T.Optional[AssignmentNode]:
+
+            # We choose one more or less arbitrary candidate
+            chosen = min(candidates2, key=lambda x: (x.lineno, x.colno))
+        elif op == 'src_add':
+            chosen = target.node
+        elif op == 'extra_files_add':
+            chosen = ArrayNode(_symbol('['), ArgumentNode(Token('', tgt_function.filename, 0, 0, 0, None, '[]')), _symbol(']'))
+
+            # this is fundamentally error prone
+            self.interpreter.dataflow_dag.add_edge(chosen, target.node)
+
+            extra_files_idnode = IdNode(Token('string', tgt_function.filename, 0, 0, 0, None, 'extra_files'))
+            if tgt_function not in self.modified_nodes:
+                self.modified_nodes += [tgt_function]
+            new_extra_files_node: BaseNode
+            if target.node.args.get_kwarg_or_default('extra_files', None) is None:
+                # Target has no extra_files kwarg, create one
+                new_kwarg_flag = True
+                new_extra_files_node = chosen
+            else:
+                new_kwarg_flag = True
+                old_extra_files = target.node.args.get_kwarg_or_default('extra_files', None)
+                target.node.args.kwargs = {k: v for k, v in target.node.args.kwargs.items() if not (isinstance(k, IdNode) and k.value == 'extra_files')}
+                new_extra_files_node = ArithmeticNode('add', old_extra_files, _symbol('+'), chosen)
+
+            tgt_function.args.kwargs[extra_files_idnode] = new_extra_files_node
+
+        newfiles_relto = self.get_relto(target.node, chosen)
+        old_src_list: T.List[T.Any] = flatten([self.interpreter.node_to_runtime_value(sn) for sn in old])
+
+        if op == 'src_add':
+            name = 'Source'
+        elif op == 'extra_files_add':
+            name = 'Extra file'
+        # Generate the new String nodes
+        to_append = []
+        added = []
+
+        old_src_list = [(target_dir_abs / x).resolve() if isinstance(x, str) else x.to_abs_path(source_root_abs) for x in old_src_list if not isinstance(x, UnknownValue)]
+        for _newf in sorted(set(newfiles)):
+            newf = Path(_newf)
+            if os.path.isabs(newf):
+                newf = Path(newf)
+            else:
+                newf = source_root_abs / newf
+            if newf in old_src_list:
+                mlog.log('  -- ', name, mlog.green(str(newf)), 'is already defined for the target --> skipping')
+                continue
+
+            mlog.log('  -- Adding ', name.lower(), mlog.green(str(newf)), 'at',
+                     mlog.yellow(f'{chosen.filename}:{chosen.lineno}'))
+            added.append(newf)
+            mocktarget = self.interpreter.funcvals[target.node]
+            assert isinstance(mocktarget, IntrospectionBuildTarget)
+            # print("adding ", str(newf), 'to', mocktarget.name) todo: should we write something to stderr?
+
+            path = str(os.path.relpath(newf, newfiles_relto))
```

Comment:
Could newf and newfiles_relto potentially be on a different drive? If so, you should use `mesonlib.relpath` to prevent exceptions...